### PR TITLE
DrivAerML: Conservative gradient clipping gc=0.5 at 4L/512d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Previous experiment (ray #2790) showed that gc=1.0 + WD=1e-2 catastrophically diverges on DrivAerML at 4L/320d (grad norms hit 231x clip). However, gradient clipping alone (without WD) has not been tested on DrivAerML. gc=0.5 provides tighter control that may stabilize training for the wider 4L/512d model without the WD-induced instability. On AirfRANS, gc=1.0 was a major breakthrough — DrivAerML may need tighter clipping given its 3D surface complexity.

## Instructions
Train on DrivAerML with 4L/512d architecture + gc=0.5 (no WD):

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent kaneda \
  --wandb-name "kaneda/drivaerml-4L512d-gc05" \
  --wandb-group "drivaerml-4L512d-gc-sweep"
```

KEY: Do NOT add weight-decay. We're testing gc alone to isolate its effect.

IMPORTANT: Must use `--batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`. Do not change these flags.

Report `val_primary/surface_rel_l2_pct` and note which epoch was best. Also report gradient norm statistics if visible in W&B.

## Baseline
Current best: **4.619%** at epoch 256 (4L/512d/8H, T_max=30, lr=5e-4, 267 epochs, 180-min)
PR: #2691 | W&B: k8qtsxxz
External target: <3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kaneda:

## Results

### DrivAerML: gc=0.5 at 4L/512d (no weight decay)

| Metric | Baseline (no gc) | gc=0.5 | Delta |
|--------|-----------------|--------|-------|
| **val_primary/surface_rel_l2_pct** | **4.619%** (ep 256) | **4.962%** (ep 257) | +7.4% (worse) |
| test_primary/surface_rel_l2_pct | N/A | 6.711% (final ep 266) | — |
| External target | <3.71% | 4.962% | 1.34x gap |

**W&B run:** [wao73xoj](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/wao73xoj)
**Epochs:** 266 (similar to baseline's 267)
**Best epoch:** 257

### Gradient norm statistics
- **Epoch 1:** grad_norm_mean=10.99, clipping 379/394 batches (96%)
- **Epoch 50:** grad_norm_mean=0.47, clipping 144/394 (37%)
- **Epoch 100:** grad_norm_mean=0.47, clipping 144/394 (37%)
- **Epoch 150:** grad_norm_mean=0.30, clipping 43/394 (11%)
- **Epoch 200:** grad_norm_mean=0.22, clipping 8/394 (2%)
- **Epoch 257 (best):** grad_norm_mean=0.16, clipping 1/394 (<1%)

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --grad-clip 0.5 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent kaneda --wandb-name "kaneda/drivaerml-4L512d-gc05" \
  --wandb-group "drivaerml-4L512d-gc-sweep"
```

### What happened

**gc=0.5 did NOT beat the no-gc baseline on DrivAerML.** Best val 4.96% vs baseline 4.62% — a 7.4% relative regression.

Key observations:
1. **Gradient clipping was highly active early** (96% of batches at epoch 1) but became almost irrelevant late in training (<1% at epoch 257). The model self-regulated its gradients over time.
2. **The trough envelope descended consistently** through 266 epochs: 61.6% → 32.5% → 12.9% → 9.3% → 7.4% → 5.4% → 5.0% → 4.96%.
3. **Tighter clipping (gc=0.5) appears to slow convergence** on DrivAerML compared to no clipping. The model reached similar epoch count (266 vs 267) but a slightly worse final metric.
4. **Unlike AirfRANS where gc was transformative**, DrivAerML's 3D surface complexity may not benefit from the conservative gradient control — the baseline optimizer already navigates the loss landscape effectively without clipping.

### Peak memory usage
Single GPU (H100 80GB), ~2.5GB VRAM used.

### Suggested follow-ups
- **gc=1.0 on DrivAerML** — looser clipping might be less restrictive while still providing some stability benefit. On AirfRANS, gc=1.0 was the initial breakthrough.
- **gc=0.5 + WD=1e-2 on DrivAerML** — the combination was powerful on AirfRANS; may transfer differently here.
- **No gc remains optimal for DrivAerML** — the baseline 4L/512d without clipping appears to be the better configuration for this dataset.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2798 because ADVISOR: Closing — your result of 4.9623% does not beat the 4.619% baseline. gc=0.5 barely triggered (consistent with taki's finding). DrivAerML needs tighter gradient clipping — we are now testing gc=0.25 and gc=0.1. Good data, kaneda..
